File: nomdl/codegen/test/struct_with_union_field_test.go
package test

import (
	"testing"

	"github.com/attic-labs/noms/nomdl/codegen/test/gen"
	"github.com/attic-labs/noms/types"
	"github.com/stretchr/testify/assert"
)

func TestStructWithUnionField(t *testing.T) {
	assert := assert.New(t)

	def := gen.StructWithUnionFieldDef{
		A: float32(1),
	}
	def = def.SetC("s")
	_, ok := def.B()
	assert.False(ok)
	c, ok := def.C()
	assert.True(ok)
	assert.Equal("s", c)

	st := def.New()
	_, ok = st.F()
	assert.False(ok)
	c, ok = st.C()
	assert.True(ok)
	assert.Equal("s", c)

	st2 := gen.NewStructWithUnionField().SetA(1).SetC("s")
	assert.True(st.Equals(st2))

	st3 := gen.NewStructWithUnionField().SetC("s").SetA(1)
	assert.True(st.Equals(st3))

	def2 := st3.Def()
	assert.Equal(def, def2)
}

func TestStructWithUnionFieldListPart(t *testing.T) {
	assert := assert.New(t)

	st := gen.NewStructWithUnionField().SetF(gen.SetOfUint8Def{2: true, 4: true}.New())
	f, ok := st.F()
	assert.True(ok)
	assert.True(f.Has(2))
	assert.False(f.Has(1))

	st2 := st.SetF(f)
	assert.True(st.Equals(st2))

	st3 := st.SetD(types.NewEmptyBlob())
	assert.False(st.Equals(st3))
	_, ok = st3.F()
	assert.False(ok)
}

File: nomdl/codegen/test/struct-with-union-field-test.js
// @flow

import {assert} from 'chai';
import {suite, test} from 'mocha';

import {StructWithUnionField} from './gen/struct_with_union_field.noms.js';

suite('struct_optional.noms', () => {
  test('constructor', async () => {
    const swuf = new StructWithUnionField({a: 1, b: 2});
    assert.equal(swuf.a, 1);
    assert.equal(swuf.b, 2);
    assert.isUndefined(swuf.c);
    assert.isUndefined(swuf.d);
    assert.isUndefined(swuf.e);
    assert.isUndefined(swuf.f);

    const swuf2 = swuf.setC('hi');
    assert.equal(swuf2.a, 1);
    assert.isUndefined(swuf2.b);
    assert.equal(swuf2.c, 'hi');
    assert.isUndefined(swuf2.d);
    assert.isUndefined(swuf2.e);
    assert.isUndefined(swuf2.f);

    assert.throws(() => {
      swuf.setC(undefined);
    });
  });
});
